ouroboros-consensus/demo-playground/Run.hs

```diff
@@ -130,15 +140,66 @@ handleSimpleNode p CLI{..} (TopologyInfo myNodeId topologyFile) = do
             }
 
       kernel <- nodeKernel nodeParams
-      let network = initNetworkLayer nodeParams kernel
+
+      let networkApp =
+            simpleSingletonVersions
+                NodeToNodeV_1
+                (NodeToNodeVersionData { networkMagic = 0 })
+                (DictVersion nodeToNodeCodecCBORTerm)
+            <$> consensusNetworkApps
+                  nullTracer
+                  nullTracer
+                  kernel
+                  ProtocolCodecs
+                    { pcChainSyncCodec =
+                        (codecChainSync
+                          (demoEncodeHeader pInfoConfig)
+                          (demoDecodeHeader pInfoConfig)
+                          (encodePoint'         pInfo)
+                          (decodePoint'         pInfo))
+                    , pcBlockFetchCodec =
+                        codecBlockFetch
+                          (demoEncodeBlock pInfoConfig)
+                          demoEncodeHeaderHash
+                          (demoDecodeBlock pInfoConfig)
+                          demoDecodeHeaderHash
+                    }
+                  (protocolHandlers nodeParams kernel)
 
       watchChain registry tracer chainDB
 
       -- Spawn the thread which listens to the mempool.
       mempoolThread <- spawnMempoolListener tracer myNodeId kernel
 
-      forM_ (producers nodeSetup) (addUpstream'   pInfo network)
-      forM_ (consumers nodeSetup) (addDownstream' pInfo network)
+      myAddr:_ <- case myNodeAddress of
+        NodeAddress host port -> getAddrInfo Nothing (Just host) (Just port)
+
+      -- TODO: cheap subscription managment, a proper one is on the way.  The
+      -- point is that it only requires 'NetworkApplications' which is a thin
+      -- layer around 'MuxApplication'.
+
+      -- serve downstream nodes
+      _ <- forkLinked registry $ do
+        versions <- runSharedState networkApp
+        (withServer myAddr (\(DictVersion _) -> acceptEq) (muxResponderNetworkApplication <$> versions) wait)
+
+      -- connect to upstream nodes
+      forM_ (producers nodeSetup) $ \na@(NodeAddress host port) ->
+        forkLinked registry $ do
+
+          let io = do
+                addr:_ <- getAddrInfo Nothing (Just host) (Just port)
+                versions <- runSharedState networkApp
+                connectTo
+                      (muxInitiatorNetworkApplication na <$> versions)
+                      -- Do not bind to a local port, use ephemeral
+                      -- one.  We cannot bind to port on which the server is
+                      -- already accepting connections.
+                      Nothing
+                      addr
+                  `catch` \(_ :: IOException) -> threadDelay 250_000 >> io
```

Comment:
Lets add a TODO here to note that this delay loop will be replaced by the connection logic later.